pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -313,13 +313,20 @@ func (a *Actuator) Exists(ctx context.Context, machine *machinev1beta1.Machine)
 	}
 	if host == nil {
 		log.Printf("Machine %v does not exist.", machine.Name)
-		return false, nil
+
+		// Clear machine addresses so that a new Node provisioned on a new Host
+		// with the same IP can be linked with its Machine and not get confused
+		// with this one.
+		return false, a.clearMachineAddresses(ctx, machine)
```

Comment:
Yeah, it feels dirty but I don't think we have an option (besides ignoring the problem) - it's working around an issue that can only really happen on baremetal and thus isn't anticipated by the Actuator interface or the Machine controller itself.

I found that I can make myself feel better by thinking about the APIs as simply methods that are called in a specific order at specific stages of the Machine's lifecycle, rather than attaching any particular meaning to them. There's nothing really special about the call to Exists(), except that it doesn't handle RequeueAfterError.